StateSpaceFlywheelSysId/robot.py

```diff
@@ -0,0 +1,115 @@
+
+import math
+import numpy
+
+# A simple utility class for converting rpm to radians, as robotPY does not currently have a wpimath.util class.
+import util.units
+
+kMotorPort = 0
+kEncoderAChannel = 0
+kEncoderBChannel = 1
+kJoystickPort = 0
+kSpinupRadPerSec = util.units.Units.rotationsPerMinuteToRadiansPerSecond(500.0)
+
+# Volts per (radian per second)
+kFlywheelKv = 0.023
+
+# Volts per (radian per second squared)
+kFlywheelKa = 0.001
+
+
+class MyRobot(wpilib.TimedRobot):
+    """
+    This is a sample program to demonstrate how to use a state-space controller to control a
+    flywheel.
+    """
+
+    def robotInit(self) -> None:
+        # The plant holds a state-space model of our flywheel. This system has the following properties:
+        #
+        # States: [velocity], in radians per second.
+        # Inputs (what we can "put in"): [voltage], in volts.
+        # Outputs (what we can measure): [velocity], in radians per second.
+        #
+        # The Kv and Ka constants are found using the FRC Characterization toolsuite.
+        self.flywheelPlant = (
+            wpimath.system.plant.LinearSystemId.identifyVelocitySystemRadians(
+                kFlywheelKv, kFlywheelKa
+            )
+        )
+
+        # The observer fuses our encoder data and voltage inputs to reject noise.
+        self.observer = wpimath.estimator.KalmanFilter_1_1_1(
+            self.flywheelPlant,
+            [3],  # How accurate we think our model is
+            [0.01],  # How accurate we think our encoder data is
+            0.020,
+        )
+
+        # A LQR uses feedback to create voltage commands.
+        self.controller = wpimath.controller.LinearQuadraticRegulator_1_1(
+            self.flywheelPlant,
+            [8],  # Velocity error tolerance
+            [12],  # Control effort (voltage) tolerance
+            0.020,
+        )
+
+        # The state-space loop combines a controller, observer, feedforward and plant for easy control.
+        self.loop = wpimath.system.LinearSystemLoop_1_1_1(
+            self.flywheelPlant, self.controller, self.observer, 12.0, 0.020
+        )
+
+        # An encoder set up to measure flywheel velocity in radians per second.
+        self.encoder = wpilib.Encoder(kEncoderAChannel, kEncoderBChannel)
+
+        self.motor = wpilib.PWMSparkMax(kMotorPort)
+
+        # A joystick to read the trigger from.
+        self.joystick = wpilib.Joystick(kJoystickPort)
+
+        # We go 2 pi radians per 4096 clicks.
+        self.encoder.setDistancePerPulse(2 * math.pi / 4096)
+
+    def teleopInit(self) -> None:
+        self.loop.reset([self.encoder.getRate()])
+
+    def teleopPeriodic(self) -> None:
+        # Sets the target speed of our flywheel. This is similar to setting the setpoint of a
+        # PID controller.
+        if self.joystick.getTriggerPressed():
+            # We just pressed the trigger, so let's set our next reference
+            self.loop.setNextR(numpy.array([kSpinUpRadPerSec]))
```

Comment:
Hm, is the numpy.array required here? I'm a bit surprised it won't accept a normal list and convert it.